jupyter-notebooks/scipy-2022-workshop/4_masks_and_filters.ipynb

```diff
@@ -0,0 +1,281 @@
+    "\n",
+    "water_mask[ndwi >= 0] = 1\n",
+    "land_mask[ndwi < 0] = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualizing Masks"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's take a look at our water and land masks:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# In the below image, purple areas are those identified as water\n",
+    "\n",
+    "plt.figure(1)\n",
+    "plt.imshow(water_mask)\n",
+    "plt.title(\"Houston water mask\")\n",
+    "plt.colorbar()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# In the below image, purple areas are those identified as land\n",
+    "\n",
+    "plt.figure(2)\n",
+    "plt.imshow(land_mask)\n",
+    "plt.title(\"Houston land mask\")\n",
+    "plt.colorbar()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You might notice that in both images, there are still small portions of water within the land areas, and vice versa. What if we wanted to clean up this mask to get a more general area of where land and water areas are? We can do this by applying morphological filters to our masks."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Apply Filters"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "More info about morphological filters using cv2: https://docs.opencv.org/4.x/d9/d61/tutorial_py_morphological_ops.html"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "closing_kernel_size = 35 #empty pixels within the mask\n",
+    "opening_kernel_size = 15 # filled pixels outside of mask"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Apply closing filter (remove empty pixels within mask)\n",
+    "closing_kernel_element = (closing_kernel_size, closing_kernel_size)\n",
+    "closing_kernel = cv2.getStructuringElement(cv2.MORPH_RECT,closing_kernel_element)\n",
+    "mask_closed = cv2.morphologyEx(np.nan_to_num(water_mask), cv2.MORPH_CLOSE, closing_kernel)\n",
+    "\n",
+    "# Apply opening filter (removing filled pixels outside of mask)\n",
+    "opening_kernel_element = (opening_kernel_size, opening_kernel_size)\n",
+    "opening_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, opening_kernel_element)\n",
+    "mask_closed_opened = cv2.morphologyEx(mask_closed, cv2.MORPH_OPEN, opening_kernel)\n",
+    "\n",
+    "# Ensure the clipped areas remain clipped\n",
+    "mask_closed_opened[mask_closed_opened == 0] = np.nan\n",
+    "\n",
+    "water_mask_filtered = mask_closed_opened"
```

Comment:
Doubling the accoladed memory for this array feels unnecessary. I'd recommend just either removing this line and replacing all instances of `water_mask_filtered` with `mask_closed_opened` OR on line 233 replacing `mask_closed_opened` with `water_mask_filtered`. That being said, it doesn't really matter. I'll yet you make the final call here. 
```suggestion
```